torch/_dynamo/_trace_wrapped_higher_order_op.py

```diff
@@ -0,0 +1,109 @@
+import functools
+
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import autograd_not_implemented
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses import FakeTensorMode
+
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+# Trace wrapped is a higher order op meant for both invoking a bound function,
+# and for registering it as a call_function in the backward graph.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the functions as needed. This, in turn, means we can support functions in backward with complex python
+# state mutation. If we were to not do this, the functions would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
+def _trace_wrapped(*args, fn):
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+_trace_wrapped_op = HigherOrderOperator("_trace_wrapped")
+
+
+def self_invoke(*args, fn):
+    # This wrapper intercepts calls to fn, and calls the real fn via _trace_wrapped
+    # Dynamo unpacks this higher order op into calling the wrapped fn
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+@_trace_wrapped_op.py_impl(ProxyTorchDispatchMode)
+def inner_trace(mode, *args, fn):
+    import torch._functorch.aot_autograd
+
+    assert len(args) == 1
+    grad = args[0]
+    assert isinstance(grad, torch.Tensor)
+
+    if isinstance(fn, functools.partial):
+        fn.__name__ = fn.func.__name__  # type: ignore[attr-defined]
```

Comment:
There's name access downstream, there are assumptions on this having a __name__ field

Reply:
You talking about in FX printing? It seems better to create a fresh function object (don't use `functools.partial`, since you *want* the `__name__` to be set).

The reason I am giving this the sideeye is you are mutating an input argument that *should* be read only. This is a violation of expectation and can lead to weird action at a distance.

Reply:
Looking

Reply:
Removed